packages/react-utilities/src/compose/types.ts

```diff
@@ -48,3 +49,54 @@ export type SlotProps<TSlots extends BaseSlots, TProps, TRootProps extends React
 } & {
   root: TRootProps;
 };
+
+/**
+ * Helper type to convert the given props of type ShorthandProps<...> into ObjectShorthandProps<...>
+ */
+export type ResolvedShorthandProps<Props, ShorthandPropNames extends keyof Props> = Omit<Props, ShorthandPropNames> &
+  { [P in ShorthandPropNames]: Props[P] extends ShorthandProps<infer T> ? ObjectShorthandProps<T> : Props[P] };
+
+/**
+ * Helper type to mark the given props as required.
+ * Similar to Required<T> except it only requires a subset of the props.
+ */
+export type RequiredProps<Props, RequiredProps extends keyof Props> = Omit<Props, RequiredProps> &
+  { [P in RequiredProps]-?: Props[P] };
+
+/**
+ * Converts a components Props type to a State type:
+ * * Adds the 'ref' and 'as' types as required values
+ * * Ensures the specified ShorthandProps are of type ObjectShorthandProps<T>
+ * * Marks the given DefaultedProps as required (-?)
+ *
+ * Example usage:
+ * ```typescript
+ * export type ExampleState = ComponentState<
+ *    ExampleProps,
+ *    { ShorthandProps: 'icon' | 'text' },
+ *    { DefaultedProps: 'color' | 'size' }
+ *  >;
+ * ```
+ */
+export type ComponentState<
+  Props extends ComponentProps,
+  ShorthandProps extends { ShorthandProps: keyof Props },
+  DefaultedProps extends { DefaultedProps: keyof Props } = { DefaultedProps: never }
+> = RequiredProps<
+  ResolvedShorthandProps<Props, ShorthandProps['ShorthandProps']>,
+  'as' | DefaultedProps['DefaultedProps']
+> & {
+  ref: React.RefObject<HTMLElement>;
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  _internal_ShorthandProps?: ShorthandProps['ShorthandProps'];
+};
+
+/**
+ * Gets the list of shorthand props from a component's State type
+ */
+export type ComponentShorthandProps<
+  State extends {
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    _internal_ShorthandProps?: string | never;
```

Comment:
It's used to squirrel away the list of ShorthandProps inside the State type, so it can be extracted later by `ComponentShorthandProps`:

```typescript
export const widgetShorthandProps: ComponentShorthandProps<WidgetState>[] = ['icon', 'text'];
```


The alternative is to make a convention that we define a separate `WidgetShorthandProps` manually, and reference that:

```typescript
export type WidgetShorthandProps = 'icon' | 'text';
export type WidgetState = ComponentState<WidgetProps, WidgetShorthandProps,
  /* DefaultedProps: */ 'text' | 'propWithDefault'
>;

export const widgetShorthandProps: WidgetShorthandProps[] = ['icon', 'text'];
```

Another alternative is to use the props array itself as the `WidgetShorthandProps` type by declaring it `as const`. This might actually be the cleanest option, because then the shorthand prop names are only ever written once (in my other proposal, they're written twice: for the `WidgetState` definition, and for `widgetShorthandProps`).
```typescript
export const widgetShorthandProps = ['icon', 'text'] as const;

export type WidgetState = ComponentState<WidgetProps, 
  /* ShorthandProps: */ typeof widgetShorthandProps[number],
  /* DefaultedProps: */ 'text' | 'propWithDefault'
>;
```

Reply:
actually the `as const` proposal sounds a lot more appealing to me, it give us one source of truth for the shorthand props with is win-win

Reply:
I've updated the proposal to use the `as const` slots array.